### ISSUE TITLE
PS4 Controller Init Bug

### ISSUE THREAD

**Issue by KelsamGames** [MonoGame/MonoGame#7457]
Originally posted here: https://community.monogame.net/t/duplicate-controller-bug/14906

I do not know if this is a bug with MonoGame, SDL, or something even deeper. Note: I've found that it is easiest to test this with a USB-connected PS4 controller, as Windows bluetooth seems to hold on to the PS4 bluetooth connection even after disconnecting for a while for some reason.

If anyone can shed some light on this, it'd be super helpful.

**If you start a MonoGame program with a PS4 controller connected and some other controller (tested with Xbox 360 and Xbox One controllers), instead of MG reporting the PS4 and Xbox controller, there are 2 instances of the Xbox controller** on 2 different GamePad indexes, and both report button presses to polling and such. If you disconnect the Xbox controller and reconnect, then the new one will work as expected, but leaves behind the "ghost" duplicate that does not report button presses or anything. If you connect the PS4 controller after the program starts (presumably, after MG/SDL have initialized), then everything works as expected, but if the programs starts with the PS4 controller and other controller connected you can see the bug. If you disconnect the PS4 controller, the "ghost" remains until you restart the program (presumably, until MG/SDL/both have re-Init-ed).

I wonder if this is a MonoGame bug, an SDL bug, or perhaps something even deeper. Seemingly not a driver-level bug, because Windows doesn't report multiple copies of the Xbox device, and this only happens if you Init the program with both connected at the same time.

**Tested with MonoGame 3.8 - DesktopGL on Windows 10 x64**

Here's a simple MG game class to test what I have described above (you will need a simple font content file to render the debug info - in this code it loads an Arial.xnb spritefont):

```
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using Microsoft.Xna.Framework.Input;
using System;
using System.Diagnostics;


namespace ControllerReader
{
	public class Game1 : Game
	{
		private GraphicsDeviceManager _graphics;
		private SpriteBatch _spriteBatch;

		string outputStr;
		SpriteFont font;

		public Game1()
		{
			_graphics = new GraphicsDeviceManager(this);
			Content.RootDirectory = "Content";
			IsMouseVisible = true;
			IsFixedTimeStep = false;
		}

		protected override void Initialize()
		{
			_graphics.PreferredBackBufferWidth = 1600;
			_graphics.PreferredBackBufferHeight = 900;
			_graphics.ApplyChanges();

			GamePad.InitDatabase();

			base.Initialize();
		}

		protected override void LoadContent()
		{
			_spriteBatch = new SpriteBatch(GraphicsDevice);
			font = Content.Load<SpriteFont>("Arial");
		}

		protected override void Update(GameTime gameTime)
		{
			for (int i = 0; i < GamePad.MaximumGamePadCount; i++)
			{
				GamePadState gamepadState = GamePad.GetState(i, GamePadDeadZone.Circular);
				GamePadCapabilities capabilities = GamePad.GetCapabilities(i);

				outputStr += $"\n{i}: ({capabilities.Identifier}) {capabilities.DisplayName}";

				// check button presses
				foreach (Buttons button in Enum.GetValues(typeof(Buttons)))
					if (gamepadState.IsButtonDown(button))
						outputStr += $" - ({(int)button}) {button}";
			}

			base.Update(gameTime);
		}

		protected override void Draw(GameTime gameTime)
		{
			GraphicsDevice.Clear(Color.Black);

			_spriteBatch.Begin();
			_spriteBatch.DrawString(font, outputStr, new Vector2(20, 20), Color.White);
			_spriteBatch.End();

			base.Draw(gameTime);
		}
	}
}
```

#### What version of MonoGame does the bug occur on:
- MonoGame 3.8 (release and develop [cloned 2021-02-03 6:19PM UTC] have been tested with the same result on both)

#### What operating system are you using:
- Windows 10 x64

#### What MonoGame platform are you using:
- DesktopGL

**Comment by mrhelmut** [comment]
Gamepad related issues, especially when involving wired PS4 controllers, are almost always related to unofficial drivers like DS4Windows.

What is your exact setup and all the gamepad related software, drivers, and virtual keys stuff like Joy2Key do you have installed? There are also remote desktop app which can totally wreck drivers like Monect.

A ghost device is typical of unofficial drivers going rogue or a third party software conflicting. Also worth noting: if you're debugging a Steam build, Steam injects its own gamepad handling into the game, entirely bypassing your own SDL bindings.

The safest route would be to test this out on a clean install of Windows with nothing but Xbox controllers. If you have a sample app and steps to reproduce, I can confirm that.

**Comment by KelsamGames** [comment]
This had actually started as troubleshooting Steam Input (still having unrelated issues with that - MAN does Steam Input not want to play nice with SDL lol), so as requested I tested on a fresh Windows 10 x64 install, and have unfortunately observed the same result.

I took a laptop that I normally use for Ubuntu testing, installed Windows 10 on it (deleting all partitions on the hard drive and formatting in the process, just to make absolutely sure it's as fresh as can be), copied over the simple MG app I pasted the game class for above, and then ran the app.

As before, if I only connect the Xbox (or multiple Xbox controllers), it behaves as expected. I can then connect the PS4 controller and it also behaves as expected. However, if I close the app and re-open it without disconnecting the PS4 controller (as long as the PS4 and at least one Xbox controller are connect when the app is opened the bug appears), then MonoGame does not show an Xbox controller and a PS4 controller connected. Rather, you will see two identical Xbox devices in indexes 0 and 1 that both respond to the single Xbox controller (among other things as described in original post). Guaranteed no Steam installed, DS4Windows, etc.

Certainly worth a try though! I was hopeful I just had something dumb going on with my system/drivers configuration upon seeing your comment.

Note: Github wouldn't let me upload the "--self-contained" build that I used, so the one attached is just a normal build published from Visual Studio (i.e. your host machine will need .NET Core 3.1)

[ControllerReader.zip](https://github.com/MonoGame/MonoGame/files/5924482/ControllerReader.zip)

**Comment by KelsamGames** [comment]
Also, for what it's worth, I rebuilt MonoGame 3.8 (Release) with the latest SDL2 (v2.0.14) on the off-chance that it is an SDL2 bug that had a fix (when cloning MonoGame 3.8, it pulls in SDL2 v2.0.12). Additionally, I tried with MonoGame's develop branch. The app being rebuilt with those versions of MonoGame produced the same bug as with standard MonoGame 3.8 (Release).

**Comment by mrhelmut** [comment]
Thanks for the testing.

I'll check if I can look into it.

For the record: Steam Input is made with SDL as well, and what it basically does with other SDL games is that it overrides the controller binding database with its own.

**Comment by KelsamGames** [comment]
Thanks for your help :) Very much appreciated!

As for Steam Input: the SDL binding db overrides I believe I have handled by clearing the "SDL_GAMECONTROLLERCONFIG" and "SDL_GAMECONTROLLER_IGNORE_DEVICES_EXCEPT" Environment variables that are set when a game is run via Steam (though any other info or a "that's not how you should do that" would be very welcome here too heh). However, even with the controller db overrides stopped, Steam Input itself still captures the input and then sends everything as their generic x-input device (if I'm not mistaken, it does this for everything but PS4 controllers that seem to actually pass through for the most part [or seemingly just have a PS4 specific db entry while everything else is just a single "x-input" GUID or some such] - weird). At any rate: it's still a big problem because we used to integrate Steam Input but decided to scrap it so we could have native support for the Xbox Series, Red Dragon Saturn, and Atari VCS Modern Controller (simply via custom SDL controller db entries). So, now, anyone that already had the game has the old Steam Input configuration that sets their controllers to send Steam Input actions and such, instead of normal input (so they just outright don't work at all). For those that didn't opt-in to Steam Input (and people that newly download the game); their controllers will get the "Generic Gamepad" template configuration and will at least work - but all appear in the game as the same device since Steam Input overrides the GUID/db that's sent to SDL with their generic x-input one (whether or not their device type is specifically opted-in to use Steam Input). On top of all that, even if I implement the piece of their API that returns the device type (from the few they can differentiate), it no longer gets accurate data and most devices return as "Unknown" (PS4 is correctly identified here, but seemingly nothing else - presumably since I'm not fully integrating Steam Input anymore and that's the only 2 that their GUIDs can differentiate). So, in the end, unless our users manually disable Steam Input for the game, the best they can hope for is a working controller - but with non-native support (detected as unknown, and therefore just assigns it basic Xbox 360 glyphs and such instead of the ones specific to the controller [like Atari VCS Modern for example, or Xbox One vs Xbox 360 vs Xbox Series]). Sorry that was pretty long-winded, but as you seem to be pretty well informed about Steam Input, I hoped maybe you had some extra info here as well that I could employ to make this end-user experience better :) Ideally, I'd like a programmatic way to just disable Steam Input entirely instead having to expect the user to figure that out (because, come on). We've even gone as far to add a little text in-game for the steam-specific build on the home/join-game screen that reads "Disable Steam Input, If Enabled", which imho is awful and ugly and basically "the game can't do this for you for some reason, so now you have to go figure this out for yourself". If we could disable Steam Input programmatically, we basically be force-dropping support for Steam Controller and Steam Link Touch entirely (which already don't work since I clear the SDL db entries Steam tries to set), but the experience playing our twin stick shooter with either of those is pretty bad anyway in our opinion. So, other than telling someone that has a Steam Controller and no other controller "sorry, you can't use that with this game", this still seems like the best option, if possible.

Upon writing this, I'm now thinking that maybe I should only override the "SDL_GAMECONTROLLER_IGNORE_DEVICES_EXCEPT" environment variable and still let the "SDL_GAMECONTROLLERCONFIG" one through, since it would (in theory) still allow the Steam Input devices to function as unknown devices while my GUIDs that Steam isn't overriding would still be natively detected. I will test with that some more and see what I find - but overall, as you can see here, this whole thing with Steam Input has been a pretty big pain point and one we still don't have a good solution for. What we're doing now is the result of days of testing and "pressing every button" just trying to get Steam Input to stop messing up our native support - so by no means am I convinced any of it is the way to go and would more than welcome any info relating to this issue as well as the primary PS4 controller bug.

**Comment by KelsamGames** [comment]
Upon re-testing after my previous comment: I have re-confirmed that anytime you run a game through steam it sets those SDL db entries (whether Steam Input is disabled for the game or not). So, I have to clear the "SDL_GAMECONTROLLERCONFIG" variable so it's entries don't override my own.

**Comment by KelsamGames** [comment]
Thinking about it, would it be possible to detect if Steam Input is "active" or not? If so, I could only override the SDL_GAMECONTROLLERCONFIG when it is "inactive" and leave it alone when "active". I could also then optionally have a thing in the game that tells the user that Steam Input is active and if they want native device support then to disable it. I haven't found any way to do this via the Steamworks SDK, but perhaps you're aware of something like an Environment variable or some such that I could utilize to accomplish this?

**Comment by mrhelmut** [comment]
I have no idea if a game can know about and wouldn't be surprised that Steam doesn't quite want us to know and prefers to enforce the use of Steam Input.

**Comment by KelsamGames** [comment]
@mrhelmut any luck with the PS4 bug? (original issue from this thread, outside of Steam)

**Comment by mrhelmut** [comment]
Unfortunately, I haven't looked into it yet.

**Comment by KelsamGames** [comment]
We are seeing this same behavior with the Nintendo Switch controller: https://community.monogame.net/t/duplicate-controller-bug/14906/3

**Comment by squarebananas** [comment]
I'm experiencing the same issue when starting with a Xbox One controller and PS4 controller connected. The Xbox One controller has a duplicate controller and no input works for the PS4 controller.


It appears the issue relates to device index values used when calling the SDL.Joystick.Open function. The device index is an unstable value which can be reassigned during startup and when disconnecting/reconnecting. Most other SDL functions use pointer values / instance ids that are stable/unique.

During startup with these 2 controllers the following events occur in the SDLRunLoop function:
- EventType.JoyDeviceAdded JoystickDevice.Which = 0
- EventType.JoyDeviceAdded JoystickDevice.Which = 0
- EventType.ControllerDeviceAdded JoystickDevice.Which = 0
- EventType.ControllerDeviceAdded JoystickDevice.Which = 1

When SDL builds it's initial list of controllers internally it can shuffle the detected controllers around and reorder them while doing so. SDL doesn't wait for the initial list to be fully built but rather each controller detected fires a JoyDeviceAdded event containing it's pre-shuffled device index before being reordered.

So when Monogame later reads the event's device index values (JoystickDevice.Which) these values are now potentially incorrect and really shouldn't be used with the SDL.Joystick.Open function. Because Monogame currently does use these values it creates duplicate controllers / has missing controllers.


There are a couple of fixes I can see to this:

1) The ControllerDeviceAdded event could be used as this seems to contain the post reordered device index value. Controllers would be added by this event instead of within Joystick.AddDevice. While this would fix Monogame's GamePad.GetState function it would leave Joystick.GetState broken and also GamePad.PacketNumber which needs Gamepads and Joysticks to be in-sync.

2) Whenever a JoyDeviceAdded event occurs attempt to open all available Joysticks (Sdl.Joystick.NumJoysticks gives the number of connected Joysticks) and don't use the received and likely out of date device index at all. Calling Sdl.Joystick.Open on an already opened Joystick simply returns it's existing pointer without issue, so checking if this already exists in Monogame's Joystick dictionary is all that is needed to prevent duplicates.


I've done a pull request for the second fix (#7486). This fix is working for me but it could do with more testing with different controller/joystick combinations.

Hope that helps and let me know if it works.

**Comment by squarebananas** [comment]
Not sure why but now the original code is working. Even though I tried SDL 2.0.14 earlier I may have not cleaned the project properly. It seems switching between 2.0.12 and 2.0.14 sends the correct device index in the event now. Interestingly 2.0.14 for me is now giving the Xbox controller's GamePadCapabilities.DisplayName as my mobile phone name, so something still isn't quite right but the GamePadState is correct.
